Expand css readme

planktonic opened the pull request:
First pass at expanding the CSS- and Sass-related parts of the readme, working towards fixing #14.

Reviewers please check:
- parts that need expanding
- sections that need editing
- is everything clear & understandable?
- tone. Any sections too dry? Any patronising?
- any 'bonus questions' to add? Should we even keep them?
- does each section flow logically from the last?
- any links to add?

All feedback more than welcome! Thanks (cc @SonniesEdge @pmeinhardt @weiland)

SonniesEdge commented:
* It's very well organised! It explains things very clearly from the ground up. 
* However, we need to consider who this is talking to. At the moment the new text read like a very basic primer in what CSS, Sass and frontend web design is, whereas previously it assumed that it was onboarding a frontend developer with some experience and giving them an overview of our naming conventions and methodologies used.
* I suggest splitting this into sections, each aimed at a particular audience. Keep the previous text as quick-intro for experienced people, use the new text for new juniors who don't know what CSS is.
* This is a public document, so the sections about our Dropbox library are incompatible with that. Either this needs to be private or those sections need to go.
  * Following from this, where is this conversation taking place? Are we talking to the public? Or new people starting at Bitcrowd?
* Bonus questions are useful if you're asking people to do research. Why are we asking these questions? Is it part of a curriculum? This again relates to who this document is talking to.

Text-specific comments are made inline.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+### Sass
+All the example CSS shown here is in the SCSS format of Sass (Syntactically Awesome Stylesheets) — this is similar to CSS (CSS is valid SCSS) but allows us more power to build & organise our styles in a DRYer and more consistent way.
+
```
Suggest explaining that Sass is compiled to CSS via asset pipeline or another build tool.

planktonic replied:
Sure, done

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+CSS was introduced to separate the presentation layer from both the data & semantics of a document, leading to easier maintenance, smaller filesizes through reuse of code, and the ability to make global layout changes with a single edit.
+
+### And everyone rejoiced!
```
Not sure if this is a reference to Monty Python or The Wiz. Either way, I would suggest using something less dependent on a specific sense of humour, which will make it more cross-culturally readable.

planktonic replied:
No reference at all, though it’s possibly a cliché. It’s merely part of this intro/history, and flows into the next heading

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+*Bonus question*: Is there an upper limit on the length of a selector?
+
+#### Still dependent on the html structure
```
HTML is capitalised.

planktonic replied:
👍  I found/replaced to all mis-capitalisations of Sass & HTML

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+Do not chain or nest selectors.
+
+*Never* style IDs (`#article-header h1`). No IDs in CSS, ever; IDs are for javascript hooks and little else. Doing either of these things starts a specificity war.
```
Overly paternalistic. Reads as shouting. Explain to the reader *why* they shouldn't do this, rather than telling them that they shouldn't.

planktonic replied:
I’m a little alarmed by 'paternalistic'. Do you mean this is sexist? (Or did you just conflate 'paternal' and 'patronising'?)

Regardless the *why* is already given, in the same paragraph.

SonniesEdge commented on README.md:
````diff
@@ -1,98 +1,348 @@
+  padding-left: 1em;
+  padding-right: 1em;
+}
+```
````
Would be good to have HTML examples alongside this. Slightly abstract without it.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [Intro to OOCSS](https://www.smashingmagazine.com/2011/12/an-introduction-to-object-oriented-css-oocss/)
+
+### Atomic design
+*Not to be confused with Atomic CSS!*
```
Again, comes across as shouting. Explain how they are different.

planktonic replied:
Yup, removing all/most/many exclamation points 👍 

An explanation of atomic css would be out of place here as we don’t use atomic css, but I’ll add a link.

SonniesEdge commented on README.md:
````diff
@@ -1,98 +1,348 @@
+  <h2 class="h5 header__subtitle">Page subtitle</h2>
+</header>
+```
+
````
It would be good to explain (with an exmplae) of how many classes together are combined together in the HTML layer to produce styles. Anyone looking up OOBEMITSCSS would find examples revolving around using `include` and `extend`, which we do not tend to do.

planktonic replied:
Yup, I hoped the example showed that — each root element of a component has the base class *and* the modifier class (I never show any different in the examples, I hope!), hence no `@extend`.

Perhaps it’s not clear enough? Let me know, but regardless — this section could do with expanding beyond one line and an example.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- `base` covers unclassed HTML elements e.g. base rules for inputs, paragraphs, images
+- `objects` are the lowest-level abstractions e.g. .button, .media, .icon
+- `components` are chunks of UI built from objects. Could be a page-type (`.profile-page`), could be a distinct block or collection (`.sidebar`, `.author-bio`). There should be no ‘reaching inside’ objects to tweak styles here (that should be achieved using --modifiers on the object directly), rather just styles to define how each object should be laid out next to the others.
+- `trumps` (named after the trump suit in card games, nothing to do with darling Donald) are high specificity overrides and helper classes e.g. `.hidden`, `.text--center`. `!important` is acceptable here.
```
As much as a dislike Donald Trump, I'd suggest leaving this out.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [Managing CSS projects with ITCSS](http://csswizardry.net/talks/2014/11/itcss-dafed.pdf) (PDF)
+
+## Using Sass well
+As with any tool, Sass enables developers to shoot themselves in the face more easily than ever before. For example the nesting functionality keeps things so nicely organised that developers initially found it tempting to go crazy and nest everything:
```
Can we leave out the references to suicide, please? And "crazy".

planktonic replied:
Yup good call (I’m still trying to wipe the slang use of 'crazy' out of my vocab. It’s sticky)

SonniesEdge commented on README.md:
````diff
@@ -1,98 +1,348 @@
+```
+The output of which breaks our carefully-constructed specificity-reducing conventions, even if the apparent encapsulation is attractive (it’s a trap!).
+
+A select few features of Sass (The Good Parts™?) can really help DRY up & organise the codebase.
````
* Not sure what "The Good Parts™" refers to.
* We need to define DRY.

planktonic replied:
Good parts refers to a super famous JS book from o’reilly. It’s famous because douglas rockford (one of the developers of JS) wrote it, it arguably lives up to its name & the hype, and because of this image:
![javascript the good parts](http://sealedabstract.com/wp-content/uploads/2013/05/JavaScript-the-good-parts.jpg)

Not important if the connection is made really, but I’ll see if I can find a suitable link to the book to make this a link at least.

First mention of DRY is now linked to the wiki page

SonniesEdge commented on README.md:
````diff
@@ -1,98 +1,348 @@
+```
+
+### Functions & Mixins
+All good developers aim for DRY code and CSS is no exception. Mixins and functions reduce the amount of CSS we manually write so reduce errors, speed up development and so on. It’s important to realise that they do not exist in the browser — the output of the mixins will be output to the final CSS every time you use them. This may seem like a waste of bits, and even very WET (so much repetition o ...
````
Define DRY and WET.

planktonic replied:
Definitions are really outside of the scope of this doc — first instance of DRY is now linked to wiki though

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [Atomic OOBEMITSCSS](http://www.sitepoint.com/atomic-oobemitscss/)
+
+### State
+There is one exception to the ‘do not nest’ rule — states. Hover, focus, visited etc. are of course all declared nested, but also any non-html-standard states we define. For our own states the naming convention is to use `is-` & `has-` as prefixes e.g. `.is-active`, `.has-child`
```
"of course" presumes previous knowledge. Explain why these are nested but other things are not.

planktonic replied:
All of this requires previous knowledge, I don’t feel we need to explain everything in depth — I’ve kept the line but asked the reader to consider *why*, in a parenthetical.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+#### Read More
+- [Atomic OOBEMITSCSS](http://www.sitepoint.com/atomic-oobemitscss/)
+
+### State
```
This should probably come after the Nesting section on #L289, as knowledge of not using nested is presumed here.

planktonic replied:
Yes definitely out of order, I’ll think about which passage to move where…

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+**DRY only refers to source files**. Repetition in source files == bad for maintainability, readability and so on. What your code looks like after any compile step is totally unrelated to this principle.
+
+### @import
+`@import` is such a reviled feature of CSS that many people don’t even know it exists (the browser makes a new, blocking request for each file `@import`ed in your main CSS file. In HTTP1 that’s a total and unthinkable performance killer). In Sass it concatenates the CSS output of the files into one, giving good performance and also allowing us to create **one CSS object or component per file**.
```
"Reviled" reads paternalistically. Let's avoid universal assumptions about these things. Better would be "We do not use @import in CSS as it makes a new blocking request..." and then follow with "In Sass it concatenates...".

planktonic replied:
Also confused by this one — what do you mean by 'paternalistic'? Is this passage sexist?

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [Chris Coyier’s CSS-Tricks](https://css-tricks.com) - Despite the cheesy name (it was created when CSS was relatively new and tricks was all anyone wanted) there are lots of great consistently-updated articles here ranging from straightforward descriptions of new CSS properties to ongoing discussions about best practices. See especially [this longform piece on SVG icons vs icon-fonts](https://c ...
+- [Brad Frost’s articles](http://bradfrost.com/blog/)
+
+## Bitcrowd Library
```
See note about where this conversation is taking place. Public or private? References to Bitcrowd library shouldn't be made in public setting.

planktonic replied:
Yes, damn security-through-obscurity dropbox links…

@xhr15: links to books in our dropbox library totally make sense for this document, but as charlie points out above, that’s a problem as this is a public document.

I see two options:
- make this repo private
- remove the links (maybe create a separate reading list repo for frontend stuff and make that private?)

SonniesEdge commented on README.md:
```diff
@@ -102,9 +352,6 @@ When using them in HTML, notice how items A and B both use the `.icon` class as
 * We use the [Autoprefixer](https://github.com/postcss/autoprefixer) postprocessor to generate browser-specific prefixes at build-time.
 
-## Browser support
-* Unless otherwise requested, we support only the latest browser versions.
```
What browsers do we support as standard?

planktonic replied:
We don’t have a standard set of browsers we support. Each client has different needs & wishes, & their projects have different requirements & target audiences. This is always done on a per-project basis. The sentence was largely meaningless as we *always* request & receive different requirements from clients, and that has never been 'only the newest'.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [CSS-Tricks — A Specificity Battle](https://css-tricks.com/a-specificity-battle/)
+
+## Further reading
+Most of the techniques & conventions described here were developed in public (Harry Roberts makes too much money from private consulting to publish things publicly) so you can still follow the threads through blog posts, and these clever people are worth reading. Find these people on Twitter, it’s still be the most useful resource for following & joining the discussion.
```
Presume "were developed in public" should be "were developed in private".

planktonic replied:
I’ve fixed the spelling & grammar (it was a horrible sentence), and then yes — cut the sentence into two. I’ve also removed the parenthetical as this doc is public.

But I definitely meant 'in public' — all these techniques get discussed and improved on a lot in the open

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+- [Nicole Sullivan’s articles](http://www.stubbornella.org/) - She started the ball rolling by creating [OOCSS](https://github.com/stubbornella/oocss) & the ubiquitous Media object.
+- [Harry Roberts’ articles](http://csswizardry.com/#section:articles) - We use some of his [Inuit.css](https://github.com/inuitcss/getting-started) objects & helpers, and ITCSS is his child. He has lots of thoughts about CSS architecture & organisation.
+- [Lea Verou](http://lea.verou.me) — awesome experiments & exploration
+- [Chris Coyier’s CSS-Tricks](https://css-tricks.com) - Despite the cheesy name (it was created when CSS was relatively new and tricks was all anyone wanted) there are lots of great consistently-updated articles here ranging from straightforward descriptions of new CSS properties to ongoing discussions about best practices. See especially [this longform piece on SVG icons vs icon-fonts](https://c ...
```
"I've seen" - doesn't match voice of rest of document.

planktonic replied:
Yes both these statements are personal opinion, as is the concept of recommending further reading — for this (and the 'read more' sections) I think it’s totally acceptable and natural-feeling to slip into a more informal tone to help indicate this switch.

…but I’ve updated the statement to make it clearer the assessment of cheesiness level is also opinion, rather than stating it so flatly.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+### A very brief history
+
+Back in the days before we had CSS, presentation was intermingled with data & semantics through such horrors as [table-based layouts](http://www.ironspider.ca/webdesign102/tables4layout2.htm#Changer3), [spacer gifs](https://en.wikipedia.org/wiki/Spacer_GIF), and visual HTML elements & attributes:
```
"such horrors" - better written as "such non-scalable techniques".

planktonic replied:
I want this to be a chatty/personable intro to how/why people have developed the techniques this doc describes, not a dry statement of a series of facts; I used the word 'horror' here for some minor amount of colour/humour.

The problems with the techniques I’ve listed here as 'horrors' are manifold, and don't include scaleability — for example table-based layouts scale just fine, it’s the lack of semantics/accessibility (& later on responsiveness) that’s the problem.

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+
+#### Everything is global
+
+There is no concept of scope in CSS so making even a small change can break other styles unexpectedly. Changing the order of your selectors can break everything, as source order is important in CSS — the latest selector overrides previous selectors of the same specificity. This feature/bug is called the Cascade, and is the ‘c’ in CSS.
```
"This feature/bug is called the Cascade" - suggest "This deliberate feature is called the Cascade".

planktonic replied:
adding 'deliberate' seems superfluous (if a feature isn't deliberate it’s a bug, no?). But I’ll make it clear somehow that it’s a feature that causes problems/is avoided/etc (that was the goal of that clause)

SonniesEdge commented on README.md:
````diff
@@ -1,98 +1,348 @@
+```
+
+Specificity starts low and ends high; earlier styles are wide-ranging, ending with very focused styles.
+- `settings` and `tools` output no CSS, they are background & setup for some sass niceness.
````
Capitalisation of "Sass".

SonniesEdge commented on README.md:
```diff
@@ -1,98 +1,348 @@
+### Functions & Mixins
+All good developers aim for DRY code and CSS is no exception. Mixins and functions reduce the amount of CSS we manually write so reduce errors, speed up development and so on. It’s important to realise that they do not exist in the browser — the output of the mixins will be output to the final CSS every time you use them. This may seem like a waste of bits, and even very WET (so much repetition o ...
+
+**GZIP is your friend**. In development mode we don’t have GZIP turned on so yes: your CSS will appear to be larger due to repeated CSS from mixin use. In production mode GZIP will be enabled, and this kind of totally consistent repetition is where GZIP excels — any extra weight from repetition will disappear.
```
Define/link GZIP. If the reader needs CSS explaining, they definitely need this.

planktonic commented:
Ok this could get confusing. I’m replying directly to inline comments, this is a reply to the not-inline/first comment. Let’s see if we know where we are at the end of it…


**Target audience**

In my experience lots of frontend devs may identify with the statement they 'can do CSS'; this most often means they understand what classes are, how to make something appear further to right/in a different colour etc. but little about CSS 'architecture':
- organisation of files, classes & selectors in order to achieve an extendable, reliable, logical & small CSS codebase
- how more general programming principles (DRY, SRP…) are being applied to CSS, why & how that helps.

This could fit the devs at our last on-site client (for example), and also our juniors.

I selected the principles I included here because they fill those holes in devs’ knowledge, and because it’s such things I most often see done poorly.


**Organisation**

Splitting into sections would be ideal, moreso as we add content. I’ve added an issue to look at that (#20) independently of this PR.


**Bonus Questions**

I added bonus questions to give direction for further learning — what might be interesting for the reader to investigate, but is out of the scope of this doc/too much detail? Same motivation with the 'read more' sections, just a different approach that may grab different people more than a reading list. No curriculum implied outside this document, and I’m certainly not expecting answers :)

planktonic commented on README.md:
```diff
@@ -1,98 +1,352 @@
+- [Brad Frost introduces Atomic design](http://bradfrost.com/blog/post/atomic-web-design/)
+- [Brad Frost writes an open book about Atomic Design](http://atomicdesign.bradfrost.com)
+
+### Atomic OOBEMITSCSS (Yes really)
```
Introduces OOBEMITSCSS before ITCSS

planktonic commented:
Ok updated, edited & expanded. Please all take another look if & when you have a moment!
